// neo4j/lib/GraphDatabase.js
/*** Generated by streamline 0.2.5 - DO NOT EDIT ***/
var __rt=require('streamline/lib/callbacks/runtime').runtime(__filename),__func=__rt.__func,__tryCatch=__rt.__tryCatch,__cb=__rt.__cb,__propagate=__rt.__propagate,__trap=__rt.__trap,__future=__rt.__future,__setEF=__rt.__setEF,__g=__rt.__g;
/*     1 */ (function() {
/*     2 */   var GraphDatabase, Node, Relationship, adjustError, status, util;
/*     4 */   status = require("http-status");
/*     6 */   util = require("./util");
/*     8 */   adjustError = util.adjustError;
/*    10 */   Relationship = require("./Relationship");
/*    12 */   Node = require("./Node");
/*    14 */   module.exports = GraphDatabase = (function() {
/*    15 */     var _this = this;
/*    17 */     GraphDatabase.name = "GraphDatabase";
/*    19 */     function GraphDatabase(url) {
/*    20 */       this.url = url;
/*    21 */       this._request = util.wrapRequestForAuth(url);
/*    22 */       this._root = null;
/*    23 */       this._services = null;
                };
/*    26 */     GraphDatabase.prototype._purgeCache = function() {
/*    27 */       this._root = null;
/*    28 */       return this._services = null;
                };
/*    31 */     GraphDatabase.prototype._getRoot = function GraphDatabase_prototype__getRoot__1(_) {
                  var response, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype__getRoot__1",
                    line: 31
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype__getRoot__1, 0, __frame, function __$GraphDatabase_prototype__getRoot__1() {
/*    33 */         if ((__this._root != null)) {
/*    34 */           return _(null, __this._root);
                    }
                  ;
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype__getRoot__1() {
/*    37 */               return __this._request.get(__this.url, __cb(_, __frame, 6, 19, function ___(__0, __1) {
/*    37 */                 response = __1;
/*    38 */                 if ((response.statusCode !== status.OK)) {
/*    39 */                   return _(response);
                            }
                          ;
/*    41 */                 __this._root = JSON.parse(response.body);
/*    42 */                 return _(null, __this._root);
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype__getRoot__1() {
                          if (error) {
/*    44 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*    48 */     GraphDatabase.prototype.getServices = function GraphDatabase_prototype_getServices__2(_) {
                  var response, root, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_getServices__2",
                    line: 48
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_getServices__2, 0, __frame, function __$GraphDatabase_prototype_getServices__2() {
/*    50 */         if ((__this._services != null)) {
/*    51 */           return _(null, __this._services);
                    }
                  ;
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getServices__2() {
/*    54 */               return __this._getRoot(__cb(_, __frame, 6, 15, function ___(__0, __1) {
/*    54 */                 root = __1;
/*    55 */                 return __this._request.get(root.data, __cb(_, __frame, 7, 19, function ___(__0, __2) {
/*    55 */                   response = __2;
/*    56 */                   if ((response.statusCode !== status.OK)) {
/*    57 */                     return _(response);
                              }
                            ;
/*    59 */                   __this._services = JSON.parse(response.body);
/*    60 */                   return _(null, __this._services);
                            }, true));
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getServices__2() {
                          if (error) {
/*    62 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*    66 */     GraphDatabase.prototype.getVersion = function GraphDatabase_prototype_getVersion__3(_) {
                  var services, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_getVersion__3",
                    line: 66
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_getVersion__3, 0, __frame, function __$GraphDatabase_prototype_getVersion__3() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getVersion__3() {
/*    69 */               return __this.getServices(__cb(_, __frame, 3, 19, function ___(__0, __1) {
/*    69 */                 services = __1;
/*    70 */                 return _(null, parseFloat((services["neo4j_version"] || "1.4")));
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getVersion__3() {
                          if (error) {
/*    72 */                 return _(adjustError);
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*    76 */     GraphDatabase.prototype.createNode = function(data) {
/*    77 */       var node;
/*    78 */       data = (data || {
                  });
/*    79 */       node = new Node(this, {
/*    80 */         data: data
                  });
/*    82 */       return node;
                };
/*    85 */     GraphDatabase.prototype.getNode = function GraphDatabase_prototype_getNode__4(url, _) {
                  var node, response, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_getNode__4",
                    line: 85
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_getNode__4, 1, __frame, function __$GraphDatabase_prototype_getNode__4() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getNode__4() {
/*    88 */               return __this._request.get(url, __cb(_, __frame, 3, 19, function ___(__0, __1) {
/*    88 */                 response = __1;
/*    89 */                 if ((response.statusCode !== status.OK)) {
/*    90 */                   if ((response.statusCode === status.NOT_FOUND)) {
/*    91 */                     return _(new Error(("No node at " + url)));
                              }
                            ;
/*    93 */                   return _(response);
                            }
                          ;
/*    95 */                 node = new Node(__this, JSON.parse(response.body));
/*    96 */                 return _(null, node);
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getNode__4() {
                          if (error) {
/*    98 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   102 */     GraphDatabase.prototype.getIndexedNode = function GraphDatabase_prototype_getIndexedNode__5(index, property, value, _) {
                  var node, nodes, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_getIndexedNode__5",
                    line: 102
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_getIndexedNode__5, 3, __frame, function __$GraphDatabase_prototype_getIndexedNode__5() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getIndexedNode__5() {
/*   105 */               return __this.getIndexedNodes(index, property, value, __cb(_, __frame, 3, 16, function ___(__0, __1) {
/*   105 */                 nodes = __1;
/*   106 */                 node = null;
/*   107 */                 if ((nodes && (nodes.length > 0))) {
/*   108 */                   node = nodes[0];
                            }
                          ;
/*   110 */                 return _(null, node);
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getIndexedNode__5() {
                          if (error) {
/*   112 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   116 */     GraphDatabase.prototype.getIndexedNodes = function GraphDatabase_prototype_getIndexedNodes__6(index, property, value, _) {
                  var key, nodeArray, nodes, response, services, url, val, _this, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_getIndexedNodes__6",
                    line: 116
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_getIndexedNodes__6, 3, __frame, function __$GraphDatabase_prototype_getIndexedNodes__6() {
                    _this = __this;
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getIndexedNodes__6() {
/*   120 */               return __this.getServices(__cb(_, __frame, 4, 19, function ___(__0, __1) {
/*   120 */                 services = __1;
/*   121 */                 key = encodeURIComponent(property);
/*   122 */                 val = encodeURIComponent(value);
/*   123 */                 url = ((((((("" + services.node_index) + "/") + index) + "/") + key) + "/") + val);
/*   124 */                 return __this._request.get(url, __cb(_, __frame, 8, 19, function ___(__0, __2) {
/*   124 */                   response = __2;
/*   125 */                   if ((response.statusCode !== status.OK)) {
/*   126 */                     return _(response);
                              }
                            ;
/*   128 */                   nodeArray = JSON.parse(response.body);
/*   129 */                   nodes = nodeArray.map(function(node) {
/*   130 */                     return new Node(_this, node);
                              });
/*   132 */                   return _(null, nodes);
                            }, true));
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getIndexedNodes__6() {
                          if (error) {
/*   134 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   138 */     GraphDatabase.prototype.getNodeById = function GraphDatabase_prototype_getNodeById__7(id, _) {
                  var node, services, url, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_getNodeById__7",
                    line: 138
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_getNodeById__7, 1, __frame, function __$GraphDatabase_prototype_getNodeById__7() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getNodeById__7() {
/*   141 */               return __this.getServices(__cb(_, __frame, 3, 19, function ___(__0, __1) {
/*   141 */                 services = __1;
/*   142 */                 url = ((("" + services.node) + "/") + id);
/*   143 */                 return __this.getNode(url, __cb(_, __frame, 5, 15, function ___(__0, __2) {
/*   143 */                   node = __2;
/*   144 */                   return _(null, node);
                            }, true));
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getNodeById__7() {
                          if (error) {
/*   146 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   150 */     GraphDatabase.prototype.createRelationship = function GraphDatabase_prototype_createRelationship__8(startNode, endNode, type, _) {
                  var __frame = {
                    name: "GraphDatabase_prototype_createRelationship__8",
                    line: 150
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_createRelationship__8, 3, __frame, _);
                };
/*   152 */     GraphDatabase.prototype.getRelationship = function GraphDatabase_prototype_getRelationship__9(url, _) {
                  var data, relationship, response, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_getRelationship__9",
                    line: 152
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_getRelationship__9, 1, __frame, function __$GraphDatabase_prototype_getRelationship__9() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getRelationship__9() {
/*   155 */               return __this._request.get(url, __cb(_, __frame, 3, 19, function ___(__0, __1) {
/*   155 */                 response = __1;
/*   156 */                 if ((response.statusCode !== status.OK)) {
/*   157 */                   return _(response);
                            }
                          ;
/*   159 */                 data = JSON.parse(response.body);
/*   160 */                 relationship = new Relationship(__this, data);
/*   161 */                 return _(null, relationship);
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getRelationship__9() {
                          if (error) {
/*   163 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   167 */     GraphDatabase.prototype.getRelationshipById = function GraphDatabase_prototype_getRelationshipById__10(id, _) {
                  var relationshipURL, services, url, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_getRelationshipById__10",
                    line: 167
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_getRelationshipById__10, 1, __frame, function __$GraphDatabase_prototype_getRelationshipById__10() {
/*   169 */         return __this.getServices(__cb(_, __frame, 2, 17, function ___(__0, __1) {
/*   169 */           services = __1;
/*   170 */           relationshipURL = services.node.replace("node", "relationship");
/*   171 */           url = ((("" + relationshipURL) + "/") + id);
/*   172 */           return __this.getRelationship(url, __cb(_, __frame, 5, 13, _, true));
                    }, true));
                  });
                };
/*   175 */     GraphDatabase.prototype.query = function GraphDatabase_prototype_query__11(query, _) {
                  var body, columns, endpoint, i, map, response, results, row, services, value, _ref, _ref1, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_query__11",
                    line: 175
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_query__11, 1, __frame, function __$GraphDatabase_prototype_query__11() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype_query__11() {
/*   178 */               return __this.getServices(__cb(_, __frame, 3, 19, function ___(__0, __2) {
/*   178 */                 services = __2;
/*   179 */                 endpoint = (services.cypher || ((((_ref = services.extensions) != null) ? (((_ref1 = _ref.CypherPlugin) != null) ? _ref1["execute_query"] : void 0) : void 0)));
/*   180 */                 if (!endpoint) {
/*   181 */                   return _(new Error("Cypher plugin not installed"));
                            }
                          ;
/*   183 */                 return __this._request.post({
/*   184 */                   uri: endpoint,
/*   185 */                   json: {
/*   186 */                     query: query
                              }
                            }, __cb(_, __frame, 8, 19, function ___(__0, __3) {
/*   183 */                   response = __3;
/*   189 */                   if ((response.statusCode === status.NO_CONTENT)) {
/*   190 */                     return _(new Error((("Unknown Neo4j error for query:\n\n" + query) + "\n")));
                              }
                            ;
/*   192 */                   if ((response.statusCode !== status.OK)) {
/*   193 */                     return _(response);
                              }
                            ;
/*   195 */                   body = response.body;
/*   196 */                   columns = body.columns;
/*   211 */                   return (function __1(_) {
                                var _i, _j, _len, _len1, _ref2, _results;
/*   199 */                     _ref2 = body.data;
/*   200 */                     _results = [];
/*   201 */                     for (_i = 0, _len = _ref2.length; (_i < _len); _i++) {
/*   202 */                       row = _ref2[_i];
/*   203 */                       map = {
                                  };
/*   204 */                       for (i = _j = 0, _len1 = row.length; (_j < _len1); i = ++_j) {
/*   205 */                         value = row[i];
/*   206 */                         map[columns[i]] = (((value && (typeof value === "object")) && value.self) ? (value.type ? new Relationship(__this, value) : new Node(__this, value)) : value);
                                  };
/*   208 */                       _results.push(map);
                                };
/*   210 */                     return _(null, _results);
/*   211 */                   })(__cb(_, __frame, 36, 18, function ___(__0, __4) {
/*   197 */                     results = __4;
/*   212 */                     return _(null, results);
                              }, true));
                            }, true));
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype_query__11() {
                          if (error) {
/*   214 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   218 */     (function(actual) {
/*   219 */       return GraphDatabase.prototype.query = function(a, b) {
/*   220 */         if ((typeof a === "function")) {
/*   221 */           console.warn((("neo4j.GraphDatabase::query()’s signature is " + "now (query, callback). Please update your code!\n") + new Error().stack.split("\n")[2]));
/*   222 */           return actual.call(this, b, a);
                    }
/*   223 */          else {
/*   224 */           return actual.apply(this, arguments);
                    }
                  ;
                  };
/*   227 */     })(GraphDatabase.prototype.query);
/*   229 */     GraphDatabase.prototype.queryNodeIndex = function GraphDatabase_prototype_queryNodeIndex__12(index, query, _) {
                  var nodeArray, nodes, response, services, url, _this, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_queryNodeIndex__12",
                    line: 229
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_queryNodeIndex__12, 2, __frame, function __$GraphDatabase_prototype_queryNodeIndex__12() {
                    _this = __this;
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype_queryNodeIndex__12() {
/*   233 */               return __this.getServices(__cb(_, __frame, 4, 19, function ___(__0, __1) {
/*   233 */                 services = __1;
/*   234 */                 url = ((((("" + services.node_index) + "/") + index) + "?query=") + (encodeURIComponent(query)));
/*   235 */                 return __this._request.get(url, __cb(_, __frame, 6, 19, function ___(__0, __2) {
/*   235 */                   response = __2;
/*   236 */                   if ((response.statusCode !== status.OK)) {
/*   237 */                     return _(response);
                              }
                            ;
/*   239 */                   nodeArray = JSON.parse(response.body);
/*   240 */                   nodes = nodeArray.map(function(node) {
/*   241 */                     return new Node(_this, node);
                              });
/*   243 */                   return _(null, nodes);
                            }, true));
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype_queryNodeIndex__12() {
                          if (error) {
/*   245 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   249 */     return GraphDatabase;
/*   251 */   }).call(this);
/*   253 */ }).call(this);
