// neo4j/lib/PropertyContainer.js
/*** Generated by streamline 0.2.5 - DO NOT EDIT ***/
var __rt=require('streamline/lib/callbacks/runtime').runtime(__filename),__defineGetter__=__rt.__defineGetter__,__defineSetter__=__rt.__defineSetter__,__func=__rt.__func,__tryCatch=__rt.__tryCatch,__cb=__rt.__cb,__propagate=__rt.__propagate,__trap=__rt.__trap,__future=__rt.__future,__setEF=__rt.__setEF,__g=__rt.__g;
/*     1 */ (function() {
/*     2 */   var PropertyContainer, adjustError, status, util;
/*     4 */   status = require("http-status");
/*     6 */   util = require("./util");
/*     8 */   adjustError = util.adjustError;
/*    10 */   module.exports = PropertyContainer = (function() {
/*    11 */     var get, set, _this = this;
/*    14 */     PropertyContainer.name = "PropertyContainer";
/*    16 */     function PropertyContainer(db, data) {
/*    17 */       this.db = db;
/*    18 */       this._request = db._request;
/*    19 */       this._data = (data || {
                  });
/*    20 */       this._data.self = ((((data != null) ? data.self : void 0)) || null);
                };
/*    23 */     get = function(props) {
/*    24 */       var getter, name, _results;
/*    25 */       _results = [];
/*    26 */       for (name in props) {
/*    27 */         getter = props[name];
/*    28 */         _results.push(PropertyContainer.prototype.__defineGetter__(name, getter));
                  };
/*    30 */       return _results;
                };
/*    33 */     set = function(props) {
/*    34 */       var name, setter, _results;
/*    35 */       _results = [];
/*    36 */       for (name in props) {
/*    37 */         setter = props[name];
/*    38 */         _results.push(PropertyContainer.prototype.__defineSetter__(name, setter));
                  };
/*    40 */       return _results;
                };
/*    43 */     get({
/*    44 */       self: function() {
/*    45 */         return (this._data.self || null);
                  }
                });
/*    49 */     get({
/*    50 */       exists: function() {
/*    51 */         return (this.self != null);
                  }
                });
/*    55 */     get({
/*    56 */       id: function() {
/*    57 */         var match;
/*    58 */         if (!this.exists) {
/*    59 */           return null;
                    }
/*    60 */          else {
/*    61 */           match = /(?:node|relationship)\/(\d+)$/.exec(this.self);
/*    62 */           return parseInt(match[1]);
                    }
                  ;
                  }
                });
/*    67 */     get({
/*    68 */       data: function() {
/*    69 */         return (this._data.data || null);
                  }
                });
/*    73 */     set({
/*    74 */       data: function(value) {
/*    75 */         return this._data.data = value;
                  }
                });
/*    79 */     PropertyContainer.prototype.equals = function(other) {
/*    80 */       return (this.self === (((other != null) ? other.self : void 0)));
                };
/*    83 */     PropertyContainer.prototype["delete"] = function PropertyContainer_prototype_delete__1(_) {
                  var message, response, __this = this;
                  var __frame = {
                    name: "PropertyContainer_prototype_delete__1",
                    line: 83
                  };
                  return __func(_, this, arguments, PropertyContainer_prototype_delete__1, 0, __frame, function __$PropertyContainer_prototype_delete__1() {
/*    85 */         if (!__this.exists) {
                      return _(null);
                    }
                  ;
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$PropertyContainer_prototype_delete__1() {
/*    89 */               return __this._request.del(__this.self, __cb(_, __frame, 6, 19, function ___(__0, __1) {
/*    89 */                 response = __1;
/*    90 */                 if ((response.statusCode !== status.NO_CONTENT)) {
/*    91 */                   message = "";
/*    92 */                   switch (response.statusCode) {
/*    93 */                   case status.NOT_FOUND:
/*    93 */                     message = "PropertyContainer not found";
/*    94 */                     break;
/*    95 */                   case status.CONFLICT:
                                message = "Node could not be deleted (still has relationships?)";
/*    96 */                   };
/*    96 */                   return _(new Error(message));
/*    97 */                 }
                          ;
                            __this._data.self = null;
/*    99 */                 __then();
                          }, true));
                        });
/*   101 */           })(function ___(error, __result) {
                        __tryCatch(_, function __$PropertyContainer_prototype_delete__1() {
                          if (error) {
                            return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
/*   103 */               }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
                return PropertyContainer;
              }).call(this);
            }).call(this);
