// neo4j/lib/Relationship.js
/*** Generated by streamline 0.2.5 - DO NOT EDIT ***/
var __rt=require('streamline/lib/callbacks/runtime').runtime(__filename),__defineGetter__=__rt.__defineGetter__,__defineSetter__=__rt.__defineSetter__,__func=__rt.__func,__tryCatch=__rt.__tryCatch,__cb=__rt.__cb,__propagate=__rt.__propagate,__trap=__rt.__trap,__future=__rt.__future,__setEF=__rt.__setEF,__g=__rt.__g;
/*     1 */ (function() {
/*     2 */   var PropertyContainer, Relationship, adjustError, status, util, __hasProp = {
/*     3 */   }.hasOwnProperty, __extends = function(child, parent) {
/*     4 */     for (var key in parent) {
/*     4 */       if (__hasProp.call(parent, key)) {
/*     4 */         child[key] = parent[key];
                  };
                };
/*     4 */     function ctor() {
/*     4 */       this.constructor = child;
                };
/*     4 */     ctor.prototype = parent.prototype;
/*     4 */     child.prototype = new ctor;
/*     4 */     child.__super__ = parent.prototype;
/*     4 */     return child;
              };
/*     6 */   status = require("http-status");
/*     8 */   util = require("./util");
/*    10 */   adjustError = util.adjustError;
/*    12 */   PropertyContainer = require("./PropertyContainer");
/*    14 */   module.exports = Relationship = (function(_super) {
/*    15 */     var get, set, _this = this;
/*    18 */     __extends(Relationship, _super);
/*    20 */     Relationship.name = "Relationship";
/*    22 */     function Relationship(db, data, start, end) {
/*    23 */       var Node;
/*    24 */       Relationship.__super__.constructor.call(this, db, data);
/*    25 */       Node = require("./Node");
/*    26 */       this._start = (start || new Node(db, {
/*    27 */         self: data.start
                  }));
/*    29 */       this._end = (end || new Node(db, {
/*    30 */         self: data.end
                  }));
                };
/*    34 */     get = function(props) {
/*    35 */       var getter, name, _results;
/*    36 */       _results = [];
/*    37 */       for (name in props) {
/*    38 */         getter = props[name];
/*    39 */         _results.push(Relationship.prototype.__defineGetter__(name, getter));
                  };
/*    41 */       return _results;
                };
/*    44 */     set = function(props) {
/*    45 */       var name, setter, _results;
/*    46 */       _results = [];
/*    47 */       for (name in props) {
/*    48 */         setter = props[name];
/*    49 */         _results.push(Relationship.prototype.__defineSetter__(name, setter));
                  };
/*    51 */       return _results;
                };
/*    54 */     get({
/*    55 */       start: function() {
/*    56 */         return (this._start || null);
                  }
                });
/*    60 */     get({
/*    61 */       end: function() {
/*    62 */         return (this._end || null);
                  }
                });
/*    66 */     get({
/*    67 */       type: function() {
/*    68 */         return this._data.type;
                  }
                });
/*    72 */     Relationship.prototype.save = function Relationship_prototype_save__1(_) {
                  var message, response, __this = this;
                  var __frame = {
                    name: "Relationship_prototype_save__1",
                    line: 72
                  };
                  return __func(_, this, arguments, Relationship_prototype_save__1, 0, __frame, function __$Relationship_prototype_save__1() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$Relationship_prototype_save__1() {
                          return (function __$Relationship_prototype_save__1(__then) {
/*    75 */                 if (__this.exists) {
/*    76 */                   return __this._request.put({
/*    77 */                     uri: (("" + __this.self) + "/properties"),
/*    78 */                     json: __this.data
                              }, __cb(_, __frame, 4, 21, function ___(__0, __1) {
/*    76 */                     response = __1;
/*    80 */                     if ((response.statusCode !== status.NO_CONTENT)) {
/*    81 */                       message = "";
/*    82 */                       switch (response.statusCode) {
/*    83 */                       case status.BAD_REQUEST:
/*    83 */                         message = "Invalid data sent";
/*    84 */                         break;
/*    85 */                       case status.NOT_FOUND:
                                    message = "Relationship not found";
/*    86 */                       };
/*    86 */                       return _(new Error(message));
/*    87 */                     }
                              ;
                                __then();
/*    89 */                   }, true));
                            }
                             else {
                              __then();
                            }
                          ;
                          })(__then);
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$Relationship_prototype_save__1() {
                          if (error) {
                            return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
/*    93 */               }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
                Relationship.prototype.del = Relationship.prototype["delete"];
                return Relationship;
              }).call(this, PropertyContainer);
            }).call(this);
